In [250]:
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split, KFold
from sklearn.preprocessing import StandardScaler
from sklearn import metrics
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn import datasets

In [251]:
base_path = r'C:\Users\frank\OneDrive\Desktop\Machine Learning\MP#4'
path = os.path.join(base_path,'vote.csv')
vote_data = pd.read_csv(path)
path2 = os.path.join(base_path,'work.csv')
work_data = pd.read_csv(path2)

# Question 1

In [252]:
vote_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   prtage    5000 non-null   int64 
 1   pesex     5000 non-null   object
 2   ptdtrace  5000 non-null   object
 3   pehspnon  5000 non-null   object
 4   prcitshp  5000 non-null   object
 5   peeduca   5000 non-null   object
 6   vote      5000 non-null   object
dtypes: int64(1), object(6)
memory usage: 273.6+ KB


In [253]:
work_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   prtage    5000 non-null   int64 
 1   pesex     5000 non-null   object
 2   ptdtrace  5000 non-null   object
 3   pehspnon  5000 non-null   object
 4   prcitshp  5000 non-null   object
 5   peeduca   5000 non-null   object
 6   work      5000 non-null   object
dtypes: int64(1), object(6)
memory usage: 273.6+ KB


# Question 2(a)

In [254]:
vote_data['vote'].unique()

array(['vote', 'did not vote'], dtype=object)

In [255]:
work_data['work'].unique()

array(['flexible', 'not flexible'], dtype=object)

In [256]:
vote_dict = {'vote':1, 'did not vote':0}
vote_data.replace({'vote':vote_dict}, inplace=True)
vote_data['vote'].unique()

array([1, 0], dtype=int64)

In [257]:
work_dict = {'flexible':1, 'not flexible':0}
work_data.replace({'work':work_dict}, inplace=True)
work_data['work'].unique()

array([1, 0], dtype=int64)

# Question 2(b)

In [258]:
vote_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   prtage    5000 non-null   int64 
 1   pesex     5000 non-null   object
 2   ptdtrace  5000 non-null   object
 3   pehspnon  5000 non-null   object
 4   prcitshp  5000 non-null   object
 5   peeduca   5000 non-null   object
 6   vote      5000 non-null   int64 
dtypes: int64(2), object(5)
memory usage: 273.6+ KB


In [259]:
work_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   prtage    5000 non-null   int64 
 1   pesex     5000 non-null   object
 2   ptdtrace  5000 non-null   object
 3   pehspnon  5000 non-null   object
 4   prcitshp  5000 non-null   object
 5   peeduca   5000 non-null   object
 6   work      5000 non-null   int64 
dtypes: int64(2), object(5)
memory usage: 273.6+ KB


There are no more discrepancies between these two data sets. 

# Question 2(c)

In [260]:
vote_data = pd.get_dummies(vote_data, columns=['pesex','ptdtrace','pehspnon','prcitshp','peeduca'])
work_data = pd.get_dummies(work_data, columns=['pesex','ptdtrace','pehspnon','prcitshp','peeduca'])

In [261]:
vote_data.head()

prtage  vote  pesex_FEMALE  pesex_MALE  ptdtrace_2 or 3 Races  \
0      19     1             1           0                      0   
1      35     1             0           1                      0   
2      48     1             0           1                      0   
3      55     0             0           1                      0   
4      25     1             1           0                      0   

   ptdtrace_American Indian, Alaskan  ptdtrace_Asian Only  ptdtrace_Asian-HP  \
0                                  0                    0                  0   
1                                  0                    0                  0   
2                                  0                    0                  0   
3                                  0                    0                  0   
4                                  0                    0                  0   

   ptdtrace_Black Only  ptdtrace_Black-AI  ...  peeduca_9TH GRADE  \
0                    0                  0  ...                  0   
1                    0                  0  ...                  0   
2                    0                  0  ...                  0   
3                    0                  0  ...                  0   
4                    0                  0  ...                  0   

   peeduca_ASSOCIATE DEGREE-ACADEMIC  peeduca_ASSOCIATE DEGREE-OCCUPATIONAL/  \
0                                  0                                       0   
1                                  0                                       0   
2                                  0                                       0   
3                                  0                                       0   
4                                  0                                       0   

   peeduca_BACHELOR'S DEGREE  peeduca_DOCTORATE DEGREE  \
0                          0                         0   
1                          0                         0   
2                          0                         0   
3                          1                         0   
4                          0                         0   

   peeduca_HIGH SCHOOL GRAD-DIPLOMA OR  peeduca_LESS THAN 1ST GRADE  \
0                                    0                            0   
1                                    0                            0   
2                                    0                            0   
3                                    0                            0   
4                                    0                            0   

   peeduca_MASTER'S DEGREE (EX: MA, MS,  peeduca_PROFESSIONAL SCHOOL DEG  \
0                                     0                                0   
1                                     1                                0   
2                                     0                                0   
3                                     0                                0   
4                                     0                                0   

   peeduca_SOME COLLEGE BUT NO DEGREE  
0                                   1  
1                                   0  
2                                   0  
3                                   0  
4                                   1  

[5 rows x 41 columns]

In [262]:
work_data.head()

prtage  work  pesex_FEMALE  pesex_MALE  ptdtrace_2 or 3 Races  \
0      35     1             0           1                      0   
1      41     1             0           1                      0   
2      53     0             0           1                      0   
3      21     1             1           0                      0   
4      45     1             1           0                      0   

   ptdtrace_4 or 5 Races  ptdtrace_American Indian, Alaskan  \
0                      0                                  0   
1                      0                                  0   
2                      0                                  0   
3                      0                                  0   
4                      0                                  0   

   ptdtrace_Asian Only  ptdtrace_Asian-HP  ptdtrace_Black Only  ...  \
0                    0                  0                    0  ...   
1                    0                  0                    0  ...   
2                    0                  0                    0  ...   
3                    0                  0                    0  ...   
4                    0                  0                    1  ...   

   peeduca_9TH GRADE  peeduca_ASSOCIATE DEGREE-ACADEMIC  \
0                  0                                  0   
1                  0                                  0   
2                  0                                  0   
3                  0                                  0   
4                  0                                  0   

   peeduca_ASSOCIATE DEGREE-OCCUPATIONAL/  peeduca_BACHELOR'S DEGREE  \
0                                       0                          0   
1                                       0                          0   
2                                       0                          1   
3                                       0                          0   
4                                       0                          0   

   peeduca_DOCTORATE DEGREE  peeduca_HIGH SCHOOL GRAD-DIPLOMA OR  \
0                         0                                    1   
1                         0                                    1   
2                         0                                    0   
3                         0                                    0   
4                         0                                    0   

   peeduca_LESS THAN 1ST GRADE  peeduca_MASTER'S DEGREE (EX: MA, MS,  \
0                            0                                     0   
1                            0                                     0   
2                            0                                     0   
3                            0                                     0   
4                            0                                     0   

   peeduca_PROFESSIONAL SCHOOL DEG  peeduca_SOME COLLEGE BUT NO DEGREE  
0                                0                                   0  
1                                0                                   0  
2                                0                                   0  
3                                0                                   1  
4                                0                                   1  

[5 rows x 40 columns]

# Question 2(d)

In [263]:
vote_data['prcitshp_FOREIGN BORN, NOT A CITIZEN OF'] = 0
vote_data['ptdtrace_4 or 5 Races'] = 0
vote_data.columns

Index(['prtage', 'vote', 'pesex_FEMALE', 'pesex_MALE', 'ptdtrace_2 or 3 Races',
       'ptdtrace_American Indian, Alaskan', 'ptdtrace_Asian Only',
       'ptdtrace_Asian-HP', 'ptdtrace_Black Only', 'ptdtrace_Black-AI',
       'ptdtrace_Black-Asian', 'ptdtrace_Hawaiian/Pacific Islander Only',
       'ptdtrace_W-A-HP', 'ptdtrace_W-B-AI', 'ptdtrace_White Only',
       'ptdtrace_White-AI', 'ptdtrace_White-Asian', 'ptdtrace_White-Black',
       'ptdtrace_White-Hawaiian', 'pehspnon_HISPANIC', 'pehspnon_NON-HISPANIC',
       'prcitshp_FOREIGN BORN, U.S. CITIZEN BY',
       'prcitshp_NATIVE, BORN ABROAD OF',
       'prcitshp_NATIVE, BORN IN PUERTO RICO OR',
       'prcitshp_NATIVE, BORN IN THE UNITED', 'peeduca_10TH GRADE',
       'peeduca_11TH GRADE', 'peeduca_12TH GRADE NO DIPLOMA',
       'peeduca_1ST, 2ND, 3RD OR 4TH GRADE', 'peeduca_5TH OR 6TH GRADE',
       'peeduca_7TH OR 8TH GRADE', 'peeduca_9TH GRADE',
       'peeduca_ASSOCIATE DEGREE-ACADEMIC',
       'peeduca_ASSOCIATE DEGREE-OCCUPA

In [264]:
work_data['ptdtrace_Black-Asian'] = 0
work_data['ptdtrace_W-A-HP'] = 0
work_data['ptdtrace_W-B-AI'] = 0 

In [265]:
work_data.columns

Index(['prtage', 'work', 'pesex_FEMALE', 'pesex_MALE', 'ptdtrace_2 or 3 Races',
       'ptdtrace_4 or 5 Races', 'ptdtrace_American Indian, Alaskan',
       'ptdtrace_Asian Only', 'ptdtrace_Asian-HP', 'ptdtrace_Black Only',
       'ptdtrace_Black-AI', 'ptdtrace_Hawaiian/Pacific Islander Only',
       'ptdtrace_White Only', 'ptdtrace_White-AI', 'ptdtrace_White-Asian',
       'ptdtrace_White-Black', 'ptdtrace_White-Hawaiian', 'pehspnon_HISPANIC',
       'pehspnon_NON-HISPANIC', 'prcitshp_FOREIGN BORN, NOT A CITIZEN OF',
       'prcitshp_FOREIGN BORN, U.S. CITIZEN BY',
       'prcitshp_NATIVE, BORN ABROAD OF',
       'prcitshp_NATIVE, BORN IN PUERTO RICO OR',
       'prcitshp_NATIVE, BORN IN THE UNITED', 'peeduca_10TH GRADE',
       'peeduca_11TH GRADE', 'peeduca_12TH GRADE NO DIPLOMA',
       'peeduca_1ST, 2ND, 3RD OR 4TH GRADE', 'peeduca_5TH OR 6TH GRADE',
       'peeduca_7TH OR 8TH GRADE', 'peeduca_9TH GRADE',
       'peeduca_ASSOCIATE DEGREE-ACADEMIC',
       'peeduca_ASSOCIATE DEGREE-

# Question 3

In [266]:
X = work_data.drop(['work'], axis=1)
y = work_data['work']

In [267]:
scaler = StandardScaler()
scaler.fit(X)
X = scaler.transform(X)

In [268]:
# Define the set of parameters to form combinations in grid search
parameters = {'kernel':('linear','poly', 'sigmoid'), 
              'C':[0.1, 1, 5, 10]}

# Run each combination using GridSearch CV
# By specifying cv = fkf, GridSearchCV will run 5-fold CV that you specified on each combination
from sklearn.model_selection import GridSearchCV
svc = SVC()
fkf = KFold(n_splits=5, random_state=26, shuffle=True)
clf = GridSearchCV(svc, parameters, cv = fkf)

clf.fit(X, y)

GridSearchCV(cv=KFold(n_splits=5, random_state=26, shuffle=True),
             estimator=SVC(),
             param_grid={'C': [0.1, 1, 5, 10],
                         'kernel': ('linear', 'poly', 'sigmoid')})

In [269]:
# Print out the mean_score for each combination
# GridSearchCV uses accuracy scores by default
for params, mean_score, rank in zip(clf.cv_results_["params"], clf.cv_results_["mean_test_score"], clf.cv_results_["rank_test_score"]):
    print(params, mean_score, rank)

{'C': 0.1, 'kernel': 'linear'} 0.8613999999999999 3
{'C': 0.1, 'kernel': 'poly'} 0.6314 12
{'C': 0.1, 'kernel': 'sigmoid'} 0.8602000000000001 5
{'C': 1, 'kernel': 'linear'} 0.8622 1
{'C': 1, 'kernel': 'poly'} 0.8352 8
{'C': 1, 'kernel': 'sigmoid'} 0.8268000000000001 9
{'C': 5, 'kernel': 'linear'} 0.8613999999999999 3
{'C': 5, 'kernel': 'poly'} 0.8516 7
{'C': 5, 'kernel': 'sigmoid'} 0.8164 10
{'C': 10, 'kernel': 'linear'} 0.8615999999999999 2
{'C': 10, 'kernel': 'poly'} 0.8532 6
{'C': 10, 'kernel': 'sigmoid'} 0.8160000000000001 11


# Question 4

The value of C that minimizes the error rate is 1 and the kernel that minimizes it is linear. 

# Question 5

In [270]:
max(clf.cv_results_['mean_test_score'])

0.8622

# Question 6

In [271]:
svc = SVC(kernel='linear',C=1.0)
svc.fit(X, y)

SVC(kernel='linear')

In [272]:
X = vote_data.drop(['vote'], axis=1)
y = vote_data['vote']

In [273]:
scaler.fit(X)
X = scaler.transform(X)

In [274]:
imputed_work = svc.predict(X)

In [275]:
vote_data['imputed_work'] = imputed_work
vote_data['imputed_work'].describe()

count    5000.000000
mean        0.643600
std         0.478983
min         0.000000
25%         0.000000
50%         1.000000
75%         1.000000
max         1.000000
Name: imputed_work, dtype: float64

# Question 7

In [276]:
import statsmodels.formula.api as smf
result = smf.ols('vote ~  imputed_work + prtage + prtage^2 + pesex_FEMALE + pesex_MALE', data = vote_data).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                   vote   R-squared:                       0.547
Model:                            OLS   Adj. R-squared:                  0.547
Method:                 Least Squares   F-statistic:                     1507.
Date:                Wed, 08 Mar 2023   Prob (F-statistic):               0.00
Time:                        23:27:34   Log-Likelihood:                -1642.9
No. Observations:                5000   AIC:                             3296.
Df Residuals:                    4995   BIC:                             3328.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        0.7174      0.022     33.296   

All of the terms were statistically significant. Of these terms all of them were positive except for age squared. The adjusted R-squared of the regression overall was 0.547 which is relatively high. 

The imputed_work coefficient is 0.2364. This means that, if all else is held equal, as the work flexibility measure increases by 1, the probability of voting increases by 23.64 percentage points. This is a relatively large increase but it must be kept in perspective that the work flexibility measure only goes from 0 to 1 so an increase of 1 is a large increase in work flexibility. 

# Question 8 

In [277]:
# First, let's define the attenuation bias correction formula
def compute_M(a,b):
    return 1 / (1 - 2 * b) * (1 - (1 - b) * b / a - (1 - b) * b / (1 - a))

In [278]:
# Next, let's calculate the inputs to the bias correction
# For the value of a, find the proportion of imputed work schedules that are flexible
a = sum(imputed_work)/(imputed_work.size)
print(a)

0.6436


In [279]:
# For the value of b, use the cross-validation error rate 
# This is left as an exercise for the mini-project, so we're just hard coding a value
b = 1 - max(clf.cv_results_['mean_test_score'])
print(b)

0.13780000000000003


In [280]:
M = compute_M(a,b)
print(M)

0.6654214712301145


# Question 9

In [281]:
work_vote_relationship = 0.2364   
work_vote_bias_correction = work_vote_relationship / M
print(work_vote_bias_correction)

0.3552635588433676


The bias corrected version is larger. And yes it does change my previous result. Because there is some noise in our prediction mixed in with the signal, correcting for this results in moving our estimates further away from zero. 